xr_fresh/dimension_reduction.py

```diff
@@ -0,0 +1,171 @@
+            np.random.choice(transposed_data.shape[0], num_samples, replace=False)
+        ]
+
+        # Fit Kernel PCA on the sampled features
+        kpca = KernelPCA(
+            kernel="rbf", gamma=gamma, n_components=n_component + 1, n_jobs=n_workers
+        )
+        kpca.fit(sampled_features)
+
+        # Extract necessary attributes from kpca for transformation
+        X_fit_ = kpca.X_fit_
+        eigenvectors = kpca.eigenvectors_[:, n_component - 1]
+        eigenvalues = kpca.eigenvalues_[n_component - 1]
+
+        @numba.jit(nopython=True, parallel=True)
+        def transform_entire_dataset_numba(
+            data, X_fit_, eigenvector, eigenvalue, gamma
+        ):
+            height, width = data.shape[1], data.shape[2]
+            transformed_data = np.zeros((height, width))
+
+            for i in numba.prange(height):
+                for j in range(width):
+                    feature_vector = data[:, i, j]
+                    k = np.exp(-gamma * np.sum((feature_vector - X_fit_) ** 2, axis=1))
+                    transformed_feature = np.dot(k, eigenvector / np.sqrt(eigenvalue))
+                    transformed_data[i, j] = transformed_feature
+
+            return transformed_data
+
+        @ray.remote
+        def process_window(
+            data_block_id,
+            data_slice,
+            window_id,
+            X_fit_,
+            eigenvector,
+            eigenvalue,
+            gamma,
+            num_workers=n_workers,
+        ):
+            data_chunk = data_block_id[
+                data_slice
+            ].data.compute()  # Convert Dask array to NumPy array
+            return transform_entire_dataset_numba(
+                data_chunk, X_fit_, eigenvector, eigenvalue, gamma
+            )
+
+        # Perform transformation in parallel
+        pt = ParallelTask(
+            self._obj,
+            row_chunks=chunk_size,
+            col_chunks=chunk_size,
+            scheduler="ray",
+            n_workers=n_workers,
+        )
+
+        # Map the process_window function to each chunk of the dataset
+        futures = pt.map(process_window, X_fit_, eigenvectors, eigenvalues, gamma)
+
+        # Combine the results
+        transformed_data = np.zeros((height, width, 1), dtype=np.float64)
+
+        # Combine the results
+        transformed_data = np.zeros((height, width))
+        for window_id, future in enumerate(ray.get(futures)):
+            window = pt.windows[window_id]
+            row_start, col_start = window.row_off, window.col_off
+            row_end, col_end = row_start + window.height, col_start + window.width
+            transformed_data[row_start:row_end, col_start:col_end] = future
+
+        # extend dimension of transformed_data
+        if len(transformed_data.shape) == 2:
+            transformed_data = np.expand_dims(transformed_data, axis=2)
+
+        # Create a new DataArray with the transformed data
+        transformed_dataarray = xr.DataArray(
```

Comment:
@jgrss Is this an easy way to get this into a format usable by gw.save() etc?  Is it just copying the attributes? I think I remember you having some helper function to copy over things like `chunksize` etc.  Also do I need to restrcture the dimensions?  At this point the output dimensions are (y,x,component#), but gw typically would do (band,y,x) does that matter?